src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AssessmentStateTest.kt

```diff
@@ -0,0 +1,169 @@
+
+        // Happy path: unallocated -> in_review -> ready_to_place -> closed
+        assessment.allocateToCurrentUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.inReview, jwt)
+
+        assessment.markReadyToPlace(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.readyToPlace, jwt)
+
+        assessment.markClosed(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.closed, jwt)
+
+        // Backwards transitions: closed -> ready_to_place -> in_review -> unallocated
+        assessment.markReadyToPlace(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.readyToPlace, jwt)
+
+        assessment.allocateToCurrentUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.inReview, jwt)
+
+        assessment.deallocateUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.unallocated, jwt)
+
+        // Unhappy path: unallocated -> in_review -> ready_to_place -> rejected
+        assessment.allocateToCurrentUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.inReview, jwt)
+
+        assessment.markReadyToPlace(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.readyToPlace, jwt)
+
+        assessment.reject(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.rejected, jwt)
+
+        // Backwards transitions: rejected -> ready_to_place -> in_review -> unallocated
+        assessment.markReadyToPlace(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.readyToPlace, jwt)
+
+        assessment.allocateToCurrentUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.inReview, jwt)
+
+        assessment.deallocateUser(jwt)
+        assessment.assertApiStatus(TemporaryAccommodationAssessmentStatus.unallocated, jwt)
+      }
+    }
+  }
+
+  private fun TemporaryAccommodationApplicationEntity.submitAndGetAssessment(jwt: String): TemporaryAccommodationAssessmentEntity {
+    webTestClient.post()
+      .uri("/applications/${this.id}/submission")
+      .header("Authorization", "Bearer $jwt")
+      .header("X-Service-Name", ServiceName.temporaryAccommodation.value)
+      .bodyValue(
+        SubmitTemporaryAccommodationApplication(
+          arrivalDate = LocalDate.now().plusDays(14),
+          type = "CAS3",
+          translatedDocument = {},
+        ),
+      )
+      .exchange()
+      .expectStatus()
+      .is2xxSuccessful
+
+    val persistedApplication = temporaryAccommodationApplicationRepository.findByIdOrNull(this.id)!!
+    return persistedApplication.assessments.first() as TemporaryAccommodationAssessmentEntity
+  }
+
+  private fun TemporaryAccommodationAssessmentEntity.assertApiStatus(status: TemporaryAccommodationAssessmentStatus, jwt: String) {
```

Comment:
Awesome, this and the other helper functions makes that very long test much more readable!